src/app/home/home-services/kmeans-local.service.ts

```diff
@@ -0,0 +1,93 @@
+import { Injectable } from '@angular/core'
+import { kmeans as KMeans } from 'ml-kmeans'
+import { ResponseInterface } from '../../interfaces/response-interface'
+
+@Injectable({
+  providedIn: 'root'
+})
+export class KmeansLocalService {
+  csvData: string[][] = []
+  clusters: any[] = []
+
+  euclideanDistance (pointA: number[], pointB: number[]): number {
+    return Math.sqrt(pointA.reduce((sum, value, index) => sum + Math.pow(value - pointB[index], 2), 0))
+  }
+
+  manhattanDistance (pointA: number[], pointB: number[]): number {
+    return pointA.reduce((sum, value, index) => sum + Math.abs(value - pointB[index]), 0)
+  }
+
+  async performKMeans (csv: File, k: number, distanceMetric: string): Promise<ResponseInterface> {
+    return await new Promise<ResponseInterface>((resolve, reject) => {
+      const fileReader = new FileReader()
+
+      fileReader.onload = () => {
+        const content = fileReader.result as string
+        this.csvData = this.csvTo2DArray(content)
+
+        const dataAsNumbers = this.csvData.slice(1)
+          .map(row => row.map(value => parseFloat(value)))
+          .filter(row => row.length === this.csvData[1].length)
+
+        const result = KMeans(dataAsNumbers, k, {
+          distanceFunction: distanceMetric === 'EUCLIDEAN' ? this.euclideanDistance : this.manhattanDistance
+        })
+
+        const jsonResult = this.convertToJSONFormat(result, dataAsNumbers, csv.name, distanceMetric)
+
+        resolve(jsonResult) // Resolviere das Promise mit dem Ergebnis
+      }
+
+      fileReader.onerror = (error) => {
+        reject(error) // Falls ein Fehler beim Lesen der Datei auftritt, rejecte das Promise
+      }
+
+      fileReader.readAsText(csv)
+    })
+  }
+
+  csvTo2DArray (csvData: string): string[][] {
```

Comment:
sollte das nicht lieber in einen eigenen Service ausgelagert werden, das kann ich dann zB. für meine Tabellenkomponente benutzen